base/iterators.jl

```diff
@@ -1431,47 +1431,64 @@ only(x::NamedTuple) = throw(
 Base.intersect(a::ProductIterator, b::ProductIterator) = ProductIterator(intersect.(a.iterators, b.iterators))
 
 """
-    IterableClosure(f)
+    Iterator(f, initialstate)
 
-An iterable object that holds a function with no parameters, `f()`, that returns an optional `Union{Some{X},Nothing}`.
+Iterable object with iterations defined by the function `f(state)`. The function takes the current state at each iteration, and follows the same
+rules as `iterate`. It must return either `(newvalue, newstate)` or `nothing`, in which case the iterator ends.
 
-`f()` is supposed to be a closure, or thunk. The iterator will produce the values from the Some objects until `nothing` is returned, signaling the end of the iterator.
+See also: [`iterate`](@ref)
 
-## Examples
+# Examples
 
 ```jldoctest
-julia> function myrange(a, b)
-           state = a
-           IterableClosure() do
-               if state <= b
-                   prevstate, state = state, state + 1
-                   Some(prevstate)
-               else
-                   nothing
-               end
+julia> fib = Iterator((1,1)) do (a,b)
+           b, (a+b, a)
```

Comment:
`a, (b, a+b)`

To keep "larger values flowing from right to left", as if they're read from left to right in ascending order? Mostly aesthetic.